Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h

```diff
@@ -0,0 +1,92 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   RecoInputContainer.h
+/// \author ole.schmidt@cern.ch
+/// \brief Struct for input data required by TRD tracking workflow
+
+#ifndef O2_TRD_RECOINPUTCONTAINER_H
+#define O2_TRD_RECOINPUTCONTAINER_H
+
+#include "ReconstructionDataFormats/TrackTPCITS.h"
+#include "DataFormatsTPC/TrackTPC.h"
+#include "DataFormatsTRD/Tracklet64.h"
+#include "DataFormatsTRD/CalibratedTracklet.h"
+#include "DataFormatsTRD/TriggerRecord.h"
+#include "DataFormatsGlobalTracking/RecoContainer.h"
+#include "CommonConstants/LHCConstants.h"
+
+#include "GPUDataTypes.h"
+
+#include <gsl/span>
+#include <memory>
+
+namespace o2
+{
+namespace trd
+{
+
+struct RecoInputContainer {
+
+  gsl::span<const o2::dataformats::TrackTPCITS> mTracksTPCITS;
+  gsl::span<const o2::tpc::TrackTPC> mTracksTPC;
+  gsl::span<const o2::trd::Tracklet64> mTracklets;
+  gsl::span<const o2::trd::CalibratedTracklet> mSpacePoints;
+  gsl::span<const o2::trd::TriggerRecord> mTriggerRecords;
+  unsigned int mNTracksTPCITS;
+  unsigned int mNTracksTPC;
+  unsigned int mNTracklets;
+  unsigned int mNSpacePoints;
+  unsigned int mNTriggerRecords;
+  std::vector<float> trdTriggerTimes;
+  std::vector<int> trdTriggerIndices;
+};
+
+inline auto getRecoInputContainer(o2::framework::ProcessingContext& pc, o2::gpu::GPUTrackingInOutPointers& ptrs, const o2::globaltracking::RecoContainer* inputTracks)
+{
+  auto retVal = std::make_unique<RecoInputContainer>();
+  retVal->mTracksTPCITS = inputTracks->getTPCITSTracks<o2::dataformats::TrackTPCITS>();
```

Comment:
I am not 100% sure whether this is allowed like this. Do we have to check which inputs we have configured, or can we just try to read them, and if the input is not there, it will return an empty span?
But even then, let's assume the input is there for whatever reason even though we have not configured it in the sources on the command line, then we should not pick them up here.

Reply:
if the input was not configured, RecoContainer will return an empty span. And since the input is configured by the command line of particular workflow, there should be no "unwanted" sources. In case you want to limit sources for matching (e.g. use only ITS-TPC, while the TPC should be anyway present in the input for the refit reason), you can see how I did this for ITS-TPC: https://github.com/shahor02/AliceO2/blob/1db020e89aa28df4f6e54d1da497b03469b3c8df/Detectors/GlobalTrackingWorkflow/src/tpcits-match-workflow.cxx#L67-L87

BTW, I see you don't use the RecoContainer::createTracks method, which accounts for redundancy of (in your case) in TPC and TPC-ITS sources. Here how I load the seeds from TPC and TPC-TOF tracks in the ITS-TPC matching: https://github.com/shahor02/AliceO2/blob/1db020e89aa28df4f6e54d1da497b03469b3c8df/Detectors/GlobalTracking/src/MatchTPCITS.cxx#L600-L629

Reply:
@shahor02 concerning the input configuration: the allowed sources (TPC,ITS-TPC) are already defined in the workflow: https://github.com/AliceO2Group/AliceO2/blob/f99e1c99b1e9777a9d29ee8146b8c97e33476df4/Detectors/TRD/workflow/src/trd-tracking-workflow.cxx#L52

The `createTracks` method I want to implement with a followup PR and get this merged first. Only a few minor things are missing now and it should be ready later today.